### PR TITLE
[feat]: [CDS-50112]: ECS task definition arn support

### PR DESCRIPTION
**Pull request by PragyeshMishra01**
## Describe your changes
Support ECS Task Definition ARN in ECS Service
Test Plan: https://docs.google.com/spreadsheets/d/1E_ZTx4NhQTFnidjjvYld1weiU1JPqK7Pl9amFaVCBdQ/edit#gid=0
## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-ALL: `trigger tiAll`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/43851)

### CODE REVIEW

**Review by sainath-b — CHANGES_REQUESTED** [review]
Lets change the approach that we discussed.
We will go ahead by creating new delegate task to handle only ARN and refactor delegate code to keep delegate handlers light weight.
cc: @rohitk-h

**Comment by PragyeshMishra01** [comment]
> Lets change the approach that we discussed. We will go ahead by creating new delegate task to handle only ARN and refactor delegate code to keep delegate handlers light weight. cc: @rohitk-h

Done

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java [hunk 1]
```diff
@@ -598,7 +598,7 @@ private ServicePartResponse executeServicePart(
     }
 
     sweepingOutputService.consume(ambiance, ServiceStepV3Constants.SERVICE_SWEEPING_OUTPUT,
-        ServiceSweepingOutput.builder().finalServiceYaml(mergedServiceYaml).build(), "");
+        ServiceSweepingOutput.builder().finalServiceYaml(mergedServiceYaml).build(), StepCategory.STAGE.name());
```
Comment (alluvamsi-harness) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java:601]: Why is this change done?

File: 125-cd-nextgen/src/main/java/io/harness/cdng/ecs/EcsStepHelperImpl.java [hunk 2]
```diff
@@ -45,9 +45,7 @@ public List<ManifestOutcome> getEcsManifestOutcome(Collection<ManifestOutcome> m
             .filter(ecsManifest -> ManifestType.EcsTaskDefinition.equals(ecsManifest.getType()))
```
Comment (alluvamsi-harness) [inline 125-cd-nextgen/src/main/java/io/harness/cdng/ecs/EcsStepHelperImpl.java:45]: This class has
if (isEmpty(ecsManifests)) {
      throw new InvalidRequestException("ECS Task Definition, Service Definition Manifests are mandatory.", USER);
}
Can we add a better message here

**Review by alluvamsi-harness — CHANGES_REQUESTED** [review]
lets also check the backward compatibility, by running automation

**Review by alluvamsi-harness — CHANGES_REQUESTED** [review]

**Comment by PragyeshMishra01** [comment]
> lets also check the backward compatibility, by running automation

I will check, but I have written it in backward compatible manner.

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java` line 606 at r5](https://reviewable.io/reviews/harness/harness-core/43851#-NONpwOo6o1S4-V8TQhd:-NONpwOo6o1S4-V8TQhe:b-f7f9wf) ([raw file](https://github.com/harness/harness-core/blob/52640bcaf36e0b7b703bb41e788214e64d31a258/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java#L606)):*
> ```Java
>     serviceCustomSweepingOutputHelper.saveAdditionalServiceFieldsToSweepingOutput(mergedServiceYaml, ambiance);
> 
>     final NGServiceV2InfoConfig ngServiceV2InfoConfig = ngServiceConfig.getNgServiceV2InfoConfig();
> ```

we have the ngServiceConfig here, so in the `saveAdditionalServiceFieldsToSweepingOutput` can we not just pass it instead of taking the yaml and then converting it to the object

**Comment by yogesh-chauhan** [comment]
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceCustomSweepingOutputHelper.java` line 65 at r5](https://reviewable.io/reviews/harness/harness-core/43851#-NONq6_z4spSpyZVwXX_:-NONq6_z4spSpyZVwXXa:b86bify) ([raw file](https://github.com/harness/harness-core/blob/52640bcaf36e0b7b703bb41e788214e64d31a258/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceCustomSweepingOutputHelper.java#L65)):*
> ```Java
>             .ecsTaskDefinitionArn(ecsServiceSpec.getEcsTaskDefinitionArn().fetchFinalValue().toString())
>             .build();
>     sweepingOutputService.consume(ambiance, ServiceStepV3Constants.SERVICE_CUSTOM_SWEEPING_OUTPUT,
> ```

Lets keep the name to be specific to ECS only instead of `ServiceStepV3Constants.SERVICE_CUSTOM_SWEEPING_OUTPUT`

File: 125-cd-nextgen/src/main/java/io/harness/cdng/ecs/EcsCanaryDeployStep.java [hunk 3]
```diff
@@ -88,24 +91,31 @@ public TaskChainResponse executeEcsTask(Ambiance ambiance, StepElementParameters
...
-            .ecsScalingPolicyManifestContentList(ecsStepExecutorParams.getEcsScalingPolicyManifestContentList())
-            .desiredCountOverride(1l)
-            .ecsServiceNameSuffix(canarySuffix)
-            .build();
-
-    TaskChainResponse response = ecsStepCommonHelper.queueEcsTask(
-        stepElementParameters, ecsCanaryDeployRequest, ambiance, executionPassThroughData, true);
+    TaskChainResponse response;
+
+    if (ecsStepExecutorParams.getEcsTaskDefinitionManifestContent() == null) {
```
[the remark this answers is not in the document]
Comment (sainath-b) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/ecs/EcsCanaryDeployStep.java:96]: Pragyesh let discuss on this conditions. We can put better conditions here.

**Review by sainath-b — APPROVED** [review]

**Review by sainath-b — DISMISSED** [review]

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java (hunk 1 above)
Comment (PragyeshMishra01) [reply 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java]: changed it now.

File: 125-cd-nextgen/src/main/java/io/harness/cdng/ecs/EcsCanaryDeployStep.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (PragyeshMishra01) [reply 125-cd-nextgen/src/main/java/io/harness/cdng/ecs/EcsCanaryDeployStep.java:96]: Done

**Comment by PragyeshMishra01** [comment]
> _[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceCustomSweepingOutputHelper.java` line 65 at r5](https://reviewable.io/reviews/harness/harness-core/43851#-NONq6_z4spSpyZVwXX_:-NONq6_z4spSpyZVwXXa:b86bify) ([raw file](https://github.com/harness/harness-core/blob/52640bcaf36e0b7b703bb41e788214e64d31a258/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceCustomSweepingOutputHelper.java#L65)):_
> 
> > ```java
> >             .ecsTaskDefinitionArn(ecsServiceSpec.getEcsTaskDefinitionArn().fetchFinalValue().toString())
> >             .build();
> >     sweepingOutputService.consume(ambiance, ServiceStepV3Constants.SERVICE_CUSTOM_SWEEPING_OUTPUT,
> > ```
> 
> Lets keep the name to be specific to ECS only instead of `ServiceStepV3Constants.SERVICE_CUSTOM_SWEEPING_OUTPUT`

Done

**Comment by PragyeshMishra01** [comment]
> _[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceCustomSweepingOutputHelper.java` line 65 at r5](https://reviewable.io/reviews/harness/harness-core/43851#-NONq6_z4spSpyZVwXX_:-NONq6_z4spSpyZVwXXa:b86bify) ([raw file](https://github.com/harness/harness-core/blob/52640bcaf36e0b7b703bb41e788214e64d31a258/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceCustomSweepingOutputHelper.java#L65)):_
> 
> > ```java
> >             .ecsTaskDefinitionArn(ecsServiceSpec.getEcsTaskDefinitionArn().fetchFinalValue().toString())
> >             .build();
> >     sweepingOutputService.consume(ambiance, ServiceStepV3Constants.SERVICE_CUSTOM_SWEEPING_OUTPUT,
> > ```
> 
> Lets keep the name to be specific to ECS only instead of `ServiceStepV3Constants.SERVICE_CUSTOM_SWEEPING_OUTPUT`

Done

**Comment by PragyeshMishra01** [comment]
> Lets change the approach that we discussed. We will go ahead by creating new delegate task to handle only ARN and refactor delegate code to keep delegate handlers light weight. cc: @rohitk-h

Done

**Comment by PragyeshMishra01** [comment]
trigger messagecheck

**Comment by PragyeshMishra01** [comment]
trigger messagecheck

**Review by alluvamsi-harness — APPROVED** [review]

**Review by sainath-b — APPROVED** [review]

**Review by sainath-b — DISMISSED** [review]
> Reviewed 2 of 12 files at r1, 22 of 35 files at r2, 1 of 2 files at r3, 2 of 6 files at r4, 2 of 4 files at r5, 17 of 17 files at r6, 9 of 9 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43851)* status: 29 file(s) to review. 1. unresolved discussions. Waiting on team(s): del-team.

**Review by alluvamsi-harness — COMMENTED** [review]
> Reviewed 1 of 12 files at r1, 19 of 35 files at r2, 1 of 6 files at r4, 2 of 4 files at r5, 5 of 17 files at r6, 1 of 9 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43851)* status: 1. unresolved discussions. Waiting on team(s): del-team.

**Review by alluvamsi-harness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43851)* status: 1. unresolved discussions. Waiting on team(s): del-team.

**Review by mdmilic — APPROVED** [review]
Approving only delegate owned files

> Reviewed 2 of 35 files at r2, 4 of 17 files at r6, 1 of 9 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43851#-:-NOkmUop4bwlGFf4N5Qk:bwc08oq)* status: 1. unresolved discussions. Waiting on team(s): del-team.

**Comment by PragyeshMishra01** [comment]
trigger ti4

**Review by sainath-b — APPROVED** [review]

**Review by sainath-b — APPROVED** [review]
> Reviewed 2 of 3 files at r8, 8 of 8 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43851)* status: 10 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained.

**Review by alluvamsi-harness — APPROVED** [review]

**Pull request merged**